README
NAME
    Device::Gsm - Perl extension to interface GSM cellular / modems

WARNING
       This is C<PRE-ALPHA> software, still needs extensive testing and
       support for custom GSM commands, so use it at your own risk,
       and without C<ANY> warranty! Have fun.

SYNOPSIS
      use Device::Gsm;

      my $gsm = new Device::Gsm( port => '/dev/ttyS1', pin => 'xxxx' );

      if( $gsm->connect() ) {
          print "connected!\n";
      } else {
          print "sorry, no connection with gsm phone on serial port!\n";
      }
 
      # Register to GSM network (you must supply PIN number in above new() call)
      $gsm->register();
 
      # Get the manufacturer and model code of device
      my $mnf   = $gsm->manufacturer();
      my $model = $gsm->model();
      print "soft version is ", $gsm->software_version(), "\n";

      my $imei = $gsm->imei() or
            $imei = $gsm->serial_number();
 
      # Test for command support
      if( $self->test_command('CGMI') ) {
          # `AT+CGMI' is supported!
      } else {
          # No luck, CGMI command not available
      }
 
      print 'Service number is now: ', $gsm->service_center(), "\n";
      $gsm->service_center( '+001505050' );   # Sets new number
  
      # Send quickly a short text message
      $modem->send_sms(
          recipient => '+3934910203040',
          content   => 'Hello world! from Device::Gsm'
      );

      # The long way...
      $modem->send_sms(

          recipient => '34910203040',
          content   => 'Hello world again, with more args',

          # SMS sending mode
          # try `text' on old phones or GSM modems
          # `pdu' is the default nowadays
          mode      => 'pdu',

          # SMS Class (can be `normal' or `flash')
          # `flash' mode delivers instantly!
          class     => 'normal'

      );
 
DESCRIPTION
    `Device::Gsm' class implements basic GSM functions, network registration
    and SMS sending.

    This class supports also `PDU' mode to send `SMS' messages, and should
    be fairly usable. I'm developing and testing it under `Linux RedHat 7.1'
    with a 16550 serial port and `Siemens C35i' / `C45' GSM phones attached
    with a Siemens-compatible serial cable.

    Please be kind to the universe and contact me if you have troubles or
    you are interested in this.

  REQUIRES

    *   Device::Modem, which in turn requires

    *   Device::SerialPort (or Win32::SerialPort on Windows machines)

  EXPORT

    None

TO-DO
    Validity Period
        Support `validity period' option on SMS sending. Tells how much time
        the SMS Service Center must hold the SMS for delivery.

    Profiles
        Build a profile of the GSM device used, so that we don't have to
        `always' test each command to know whether it is supported or not,
        because this takes too time to be done every time.

AUTHOR
    Cosimo Streppone, cosimo@cpan.org

SEE ALSO
    the Device::Modem manpage, the Device::SerialPort manpage, the
    Win32::SerialPort manpage, perl(1)



Changes
Revision history for Perl extension Device::Gsm.

1.49  Sat Aug  9 15:12:00 CET 2008
    - Modified manufacturer() to work also with some Motorola phones,
      that report AT+CGMI in a slightly different way.
    - Changed by phone number :)

1.48  Wed Feb 28 21:55:23 CET 2007
    - Fixed CPAN bug #24781, thanks to torsten at archesoft dot de for reporting.

1.47  Tue Feb 13 11:31:24 CET 2007
    - Fixed t/30networks.t test script.

1.46  Tue Dec 19 22:05:31 CET 2006
    - Fixed CPAN RT wishlist #23575.
      Again thanks to Troels for his ideas.

1.45  Tue Oct 17 18:01:43 CEST 2006
    - Fixed CPAN RT bugs #21991, #21992, #22025.
      Thanks to Troels Jensen for accurate reporting and patch suggestions.

1.44  Wed Aug 16 08:25:50 CEST 2006
    - Fixed decoding of alphabetical sender addresses.
    - Added correct decoding of UCS languages (Russian for example).
      Thanks to Nikolay Shaplov for the bug report + patch.

1.43  Sun Jul 23 17:40:15 CEST 2006
    - Added ability to decode alphabetical sender addresses.
      Thanks to Torsten Mueller for report the issue.

1.42  Mon May  8 21:16:40 CEST 2006
    - Fixed a bug in Device::Gsm::Sms delete() method that invoked
      a non-existent Device::Gsm::deleteMessage().
    - Implemented read and delete of sms messages from different
      storages (ME, SM, ...). Only for PDU format. Thanks to
      Jurij Sikorsky for the patch.

1.41  Thu Apr 20 21:52:19 CEST 2006
    - === Now Device::Gsm 1.39+ requires at least Device::Modem v1.47 ===
      Check out on CPAN before upgrading.
    - Some cool fixes for better support of Iridium Satellite phones,
      GPRS connections and command/response faster cycle.
      Many thanks to Ed Wildgoose.

1.37  Sat Aug 27 12:53:00 CEST 2005
    - fixed regexp to extract sim card sms messages information
      (thanks to Pierre Hilson <zorglups hotmail com>)

1.36
1.35  Wed Sep 15 23:43:03 CEST 2004
    - finally added encoding/decoding of sms text between
      8 bit ASCII (ISO-8859) and 7 bit GSM 03.38
      (thanks to Stefano Pisani for his feedback)

1.34  Wed Aug 18 09:10:48 CEST 2004
    - fixed delete_sms() command syntax and results parsing. Thanks
      to all users that reported problems.
    - added an example script on how to delete sms messages.

1.33  Wed May 26 13:52:43 CEST 2004
    - added delete_sms() method. For now, works only in PDU mode.
    - fixed messages() newline regex split pattern
    - modified send sms in text mode to cope with slower/older devices
    - modified message read method to set pdu mode before executing

1.28
  ||
1.25  Fri Jan 23 00:59:16 CET 2004
    - further documentation, troubleshooting and autoscan utility
	  sections
	- added all new pdu sms token-oriented decoding classes
	  (this is an 80% complete implementation, mostly regarding charset
	  conversion issues)
	- Device::Gsm::Sms public class is now documented properly

1.24
  ||
1.20  Wed Jan 21 23:32:45 CET 2004
	- completed full documentation (to be revised)
	- documented working irda/bluetooth connections (always
	  through device::serialport :-)

1.19
1.18  Mon Mar 24 23:32:27 CET 2003
	- added a complete interface to read messages on SIM card
	  ( $gsm->messages() method and Device::Gsm::Sms objects )
	- some minor cleanup

1.17  Thu Sep 12 00:22:06 CEST 2002
	- added signal_quality() method to measure gsm signal power (dBm)
	- added an example script in [examples] folder (`report_signal.pl')

1.15
1.16  Tue Sep  3 23:37:56 CEST 2002
	- fixed documentation example (thanks to Domenico Viggiani)
	- added `examples/send_to_cosimo.pl' script to report successful
	  installations of Device::Gsm module :-)
	- improved test scripts; now work also with env vars, and without too
	  many `connect()'s

1.14  Mon Apr 29 18:54:36 CEST 2002
	- fixed errors in debugging
	- fixed errors in AT command response handling (still to-do)
	- included network registration test (must supply PIN in t/30register.t)
	- added send_sms.pl in examples/

1.10  Wed Apr 10 00:20:28 CEST 2002
	- fixed and renamed Pdu.pm functions
	- send_sms() new parameters that dispatch calls to
	  _send_sms_pdu() and _send_sms_text() to allow text and PDU modes
	- send_sms() now works! (with siemens C25/35/45)

1.09  Tue Apr  9 00:19:02 CEST 2002
	- Pdu.pm added with functions to treat PDU encoded data
	- added _send_sms_pdu() internal method

1.08  Sat Apr  6 00:24:29 CEST 2002
	- hangup() method
	- imei() / serial_number() method with test case

1.07  Wed Apr  3 23:38:51 CEST 2002
	- removed unnecessary javadoc style tags
	- made use of new parse_answer() call
	- now service_center() works
	- device identification functions now work
	- improved test suite
	- fixed logging class mechanism

1.06  Sat Mar 30 16:43:55 CET 2002
	- added manufacturer(), model(), software_version() methods to 
      get information about the device
	- added test_command() to test support for a specific GSM command

1.05  Sat Mar 30 15:12:09 CET 2002
	- added service_number() method

1.04  Mon Mar 25 07:34:52 CET 2002
    - added pod, html, text docs

0.01  Tue Feb 26 22:15:31 CET 2002
	- original version; created by h2xs 1.20 with options
		-A -X -n Device::Gsm -v 0.01




lib/Device/Gsm/Sms.pm
# Device::Gsm::Sms - SMS message simple class that represents a text SMS message
# Copyright (C) 2002-2006 Cosimo Streppone, cosimo@cpan.org
#
# This program is free software; you can redistribute it and/or modify
# it only under the terms of Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# Perl licensing terms for details.
#
# Commercial support is available. Write me if you are
# interested in new features or software support.
#
# $Id: Sms.pm,v 1.14 2006/10/17 15:53:18 cosimo Exp $

package Device::Gsm::Sms;
$Device::Gsm::Sms::VERSION = substr q$Revision: 1.14 $, 10;

use strict;
use integer;

use constant SMS_DELIVER => 0x00;
use constant SMS_SUBMIT  => 0x01;
use constant SMS_STATUS  => 0x02;

use Device::Gsm::Pdu;
use Device::Gsm::Sms::Structure;
use Device::Gsm::Sms::Token;

sub _log    { print @_, "\n"; }
sub _parent { $_[0]->{_parent} }

#
# new(
#     header  => '+CMGL: .....',
#     pdu     => '[encoded pdu string]',
# )
#
# creates message object
#
sub new {
	my($proto, %opt) = @_;
	my $class = ref $proto || $proto;

	# Create new message object
	my $self = {};

    # Store gsm parent object reference
    if( exists $opt{'parent'} ) {
        $self->{'_parent'} = $opt{'parent'};
        # Assume default storage for sms message
        $opt{'storage'} ||= $self->{'_parent'}->storage();
    }

    # Store options into main object
	$self->{'options'} = \%opt;

	# Hash to contain token objects after decoding (must be accessible by name)
	$self->{'tokens'}  = {};

	return undef unless( exists $opt{'header'} && exists $opt{'pdu'} );

#_log("NEW SMS OBJECT");
#_log("Header [$opt{header}]");
#_log("PDU    [$opt{pdu}]");

	# Check for valid msg header (thanks to Pierre Hilson for his patch
	# to make this regex work also for Alcatel gsm software)
	if( $opt{'header'} =~ /\+CMGL:\s*(\d+),\s*(\d+),\s*(\w*),\s*(\d+)/o )
	{

		$self->{'index'}  = $1;                        # Position of message in SIM card
		$self->{'status'} = $2;                        # Status of message (REC READ/UNREAD, STO, ...);
		$self->{'alpha'}  = $3;                        # Alphanumeric representation of sender
		$self->{'length'} = $4;                        # Final length of message
		$self->{'pdu'}    = $opt{'pdu'};               # PDU content
		$self->{'storage'}= $opt{'storage'};           # Storage (SM or ME)

		bless $self, $class;

		if( $self->decode( Device::Gsm::Sms::SMS_DELIVER ) ) {
#			_log('OK, message decoded correctly!');
		} else {
#			_log('CASINO!');
			undef $self;
		}

	} else {

		# Warning: could not parse message header
		undef $self;

	}

	return $self;
}

#
# time(): returns message time in ascii format
#
sub time {
	my $self = shift;
	if( my $t = $self->token('SCTS') ) {
		return $t->toString();
	}
	return '';
}

#
# type(): returns message type in ascii readable format
#
{
	# List of allowed status strings
	my @status = ( 'UNKNOWN', 'REC UNREAD', 'REC READ', 'SENT UNREAD', 'SENT READ' );

	sub status () {
		my $self = shift;
		return $status[ defined $self->{'status'} ? $self->{'status'} : 0 ];
	}

}

#
# decode( CMGL_header, pdu_string )
#
# creates a new Device::Gsm::Sms object from
# PDU encoded message string returned by +CMGL commands
#
# If some error occurs, returns undef.
#
#
sub _old_decode {
	my($header, $pdu) = @_;
	my %msg = ();
	my $errors = 0;

	# Copy original header/pdu strings
	$msg{'_HEADER'} = $header;
	$msg{'_PDU'} = $pdu;

	#
	# Decode header string
	#
	if( $header =~ /\+CMGL:\s*(\d+),(\d+),(\d*),(\d+)/ ) {
		$msg{'index'}  = $1;
		$msg{'type'}   = $2;
		$msg{'xxx'}    = $3;   # XXX
		$msg{'length'} = $4;
	}

	#
	# Decode all parts of PDU message
	#

	# ----------------------------------- SCA (service center address)
	my $sca_length = hex( substr $pdu, 0, 2 );
	if( $sca_length == 0 ) {
		# No SCA provided, take default
		$msg{'SCA'} = undef;
	} else {
		# Parse SCA address
		#print STDERR "SCA length = ", $sca_length, "; ";
		#print STDERR "Parsing address ", substr( $pdu, 0, ($sca_length+1) << 1 );
		$msg{'SCA'} = Device::Gsm::Pdu::decode_address( substr($pdu, 0, ($sca_length+1) << 1 ) );
		#print STDERR ' = `', $msg{'SCA'}, "'\n";
	}

	# ----------------------------------- PDU type
	$pdu = substr $pdu => (($sca_length+1) << 1);
	$msg{'PDU_TYPE'} = substr $pdu, 0, 2;
	undef $sca_length;

	# ----------------------------------- OA (originating address)
	$pdu = substr $pdu => 2;
	my $oa_length = hex( substr $pdu, 0, 2 );

	$msg{'OA'} = Device::Gsm::Pdu::decode_address( substr($pdu, 0, ($oa_length+1) << 1 ) );
	undef $oa_length;

	# PID      (protocol identifier)
	# DCS      (data coding scheme)
	# SCTS     (service center time stamp)
	# UDL + UD (user data)
	@msg{ qw/PID DCS SCTS UDL UD/ } = unpack 'A2 A2 A14 A2 A*', $pdu;

	#map { $msg{$_} = hex $msg{$_} } qw/PID DCS UDL/;
	#
	# Decode USER DATA in 7/8 bit encoding
	#
	if( $msg{'DCS'} eq '00' ) { # DCS_7BIT
		Device::Gsm::Pdu::decode_text7( $msg{'UD'} );
	} elsif( $msg{'DCS'} eq 'F6' ) { # DCS_8BIT
		Device::Gsm::Pdu::decode_text8( $msg{'UD'} );
	}

	# XXX DEBUG
	#foreach( sort keys %msg ) {
	#	print STDERR 'MSG[', $_, '] = `'.$msg{$_}.'\'', "\n";
	#}

	bless \%msg, 'Device::Gsm::Sms';
}


sub decode {
	my( $self, $type ) = @_;
	$self->{'type'} = $type;

	# Get list of tokens for this message (from ::Sms::Structure)
	my $cPdu        = $self->{'pdu'};

	# Check that PDU is not empty
	return 0 unless $cPdu;

	# Backup copy for "backtracking"
	my $cPduCopy    = $cPdu;

	my @token_names = $self->structure();
	my $decoded     = 1;

	while( @token_names ) {

		# Create new token object
		my $token = new Sms::Token( shift @token_names, {messageTokens => $self->{'tokens'}} );
		if( ! defined $token ) {
			$decoded = 0;
			last;
		}

		# If decoding is completed successfully, add token object to message
#_log('PDU BEFORE ['.$cPdu.']', length($cPdu) );

		if( $token->decode(\$cPdu) ) {

			# Store token object into SMS message
			$self->{'tokens'}->{ $token->name() } = $token;

			# Catch message type indicator (MTI) and re-load structure
            # We must also skip message types 0x02 and 0x03 because we don't handle them currently
			if( $token->name() eq 'PDUTYPE' ) {
            
                my $mti = $token->MTI();

=cut
                # If MTI has bit 1 on, this could be a SMS-STATUS message (0x02), or (0x03???)
                if( $mti >= SMS_STATUS ) {
                    _log('skipping unhandled message type ['.$mti.']');
                    return undef;
                }
=cut

                if( $mti != $type ) {
#_log('token PDUTYPE, data='.$token->data().' MTI='.$token->get('MTI').' ->MTI()='.$token->MTI());
                    #
                    # This is a SMS-SUBMIT message, so:
                    #
                    # 1) change type
                    # 2) restore original PDU message
                    # 3) reload token structure
                    # 4) restart decoding
                    #
                    $self->type( $type = $mti );

                    $cPdu = $cPduCopy;
                    @token_names = $self->structure();

#_log('RESTARTING DECODING AFTER MTI DETECTION'); #<STDIN>;
    				redo;
	    		}

#_log('       ', $token->name(), ' DATA = ', $token->toString() );

            }

		}

#_log('PDU AFTER  ['.$cPdu.']', length($cPdu) );

	}

#_log("\n", 'PRESS ENTER TO CONTINUE'); <STDIN>;

	return $decoded;

}

#
# Delete an sms message
#
sub delete {
    my $self = $_[0];
    my $gsm  = $self->_parent();
    my $ok;

    # Try to delete message
    my $msg_index = $self->index();
    my $storage = $self->storage();

    # Issue delete command
    if( ref $gsm && $msg_index > 0 ) {
        $ok = $gsm->delete_sms($msg_index, $storage);
        $gsm->log->write('info', 'Delete sms n.'.$msg_index.' in storage '.$storage.' => '.($ok?'OK':'*ERROR'));
    } else {
        $gsm->log->write('warn', 'Could not delete sms n.'.$msg_index.' in storage '.$storage.'. Internal error.');
        $ok = undef;
    }

    return $ok;
}

#
# Returns message own index number (position) 
#
sub index {
    my $self = $_[0];
    return $self->{'index'};
}

#
# Returns message storage (SM - SIM card or ME - phone memory)
#
sub storage {
    my $self = $_[0];
    return $self->{'storage'};
}

#
# Only valid for SMS_SUBMIT messages (?)
#
sub recipient {
	my $self = shift;
	if( $self->type() == SMS_SUBMIT ) {
		my $t = $self->token('DA');
		return $t->toString() if $t;
	}
}

#
# Only valid for SMS_DELIVER messages (?)
#
sub sender {
	my $self = shift;
	if( $self->type() == SMS_DELIVER ) {
		my $t = $self->token('OA');
		return $t->toString() if $t;
	}
}

# Alias for text()
sub content {
	return $_[0]->text();
}

sub text {
	my $self = shift;
	my $t = $self->token('UD');
	return $t->toString() if $t;
}

sub token ($) {
	my($self, $token_name) = @_;
	return undef unless $token_name;

	if( exists $self->{'tokens'}->{$token_name} ) {
		return $self->{'tokens'}->{$token_name};
	} else {
		warn('undefined token '.$token_name.' for this sms');
		return undef;
	}
}

#
# Returns type of sms (SMS_DELIVER || SMS_SUBMIT)
#
sub type {
	my $self = shift;
	if( @_ ) {
		$self->{'type'} = shift;
	}
	$self->{'type'};
}



=pod

=head1 NAME

Device::Gsm::Sms - SMS message internal class that represents a single text SMS message

=head1 SYNOPSIS

    # A list of Device::Gsm::Sms messages is returned by
    # Device::Gsm messages() method.

    use Device::Gsm;
    ...
    @sms = $gsm->messages();

    if( @sms ) {
        foreach( @sms ) {
            print $msg->storage()   , "\n";
            print $msg->recipient() , "\n";
            print $msg->sender()    , "\n";
            print $msg->content()   , "\n";
            print $msg->time()      , "\n";
            print $msg->type()      , "\n";
        }
    }

    # Or you can instance a sms message from raw PDU data
    my $msg = new Device::Gsm::Sms(
        header => '+CMGL: ...',
        pdu    => `[encoded pdu data]',
        storage=> 'ME', # or 'SC'
    );

    if( defined $msg ) {
        print $msg->recipient() , "\n";
        print $msg->sender()    , "\n";
        print $msg->content()   , "\n";  # or $msg->text()
        print $msg->time()      , "\n";
        print $msg->type()      , "\n";
    }

=head1 DESCRIPTION

C<Device::Gsm::Sms> class implements very basic SMS message object,
that can be used to decode C<+CMGL> GSM command response to build a more
friendly high-level object.

=head1 METHODS

The following is a list of methods applicable to C<Device::Gsm::Sms> objects.

=head2 content()

See text() method.

=head2 decode()

Starts the decoding process of pdu binary data. If decoding process 
ends in success, return value is true and sms object is filled with
all proper values.

If decoding process has errors or pdu data is not provided, return
value is 0 (zero).


=head2 delete()

Delete the current SMS message from sim card.
Example:

    $gsm = Device::Gsm->new();
    ...
    my @msg = $gsm->messages();
    $msg[0] && $msg[0]->delete();

=head2 new()

Basic constructor. You can build a new C<Device::Gsm::Sms> object from the
raw B<+CMGL> header and B<PDU> data. Those data is then decoded and a new
sms object is instanced and all information filled, to be available
for subsequent method calls.

The allowed parameters to new() method are:

=over 4

=item header

This is the raw B<+CMGL> header string as modem outputs when you
issue a B<+CMGL> command

=item pdu

Binary encoded sms data

=item storage

Tells which storage to delete the message from. Check the documentation of your
phone to know valid storage values. Default values are:

=over 4

=item C<ME>

Deletes messages from gsm phone memory.

=item C<SC>

Deletes messages from sim card.

=back

=back

=head2 index()

Returns the sms message index number, that is the position of message in the
internal device memory or sim card.
This number is used for example to delete the message.

    my $gsm = Device::Gsm->new(port=>'/dev/ttyS0');
    ...
    my @messages = $gsm->messages();
    ...
    # Delete the first returned message
    my $msg = shift @messages;
    $gsm->delete_sms( $msg->index() );

=head2 recipient()

Returns the sms recipient number (destination address = DA)
as string (ex.: C<+39012345678>).

=head2 sender()

Returns the sms sender number (originating address = OA) as string.

=head2 status()

Status of the message can be one value from the following list:

=for html
<FORM><SELECT><OPTION>UNKNOWN<OPTION>REC UNREAD<OPTION>REC READ<OPTION>SENT UNREAD<OPTION>SENT READ</SELECT></FORM>

=for pod
'UNKNOWN', 'REC UNREAD', 'REC READ', 'SENT UNREAD', 'SENT READ'

=head2 storage()

Returns the storage where SMS has been read from.

=head2 text()

Returns the textual content of sms message.

=head2 token()

Returns the given PDU token of the decoded message (internal usage).

=head2 type()

SMS messages can be of two types: SMS_SUBMIT and SMS_DELIVER, that are defined by
two constants with those names. type() method returns one of these two values.

Example:

	if( $sms->type() == Device::Gsm::Sms::SMS_DELIVER ) {
		# ...
	}
	elsif( $sms->type() == Device::Gsm::Sms::SMS_SUBMIT ) {
		# ...
	}

=head1 REQUIRES

=over 4

=item *

Device::Gsm

=back

=head1 EXPORTS

None

=head1 TODO

=over 4

=item *

Complete and proof-read documentation and examples

=back

=head1 COPYRIGHT

Device::Gsm::Sms - SMS message simple class that represents a text SMS message

Copyright (C) 2002-2006 Cosimo Streppone, cosimo@cpan.org

This program is free software; you can redistribute it and/or modify
it only under the terms of Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
Perl licensing terms for details.

=head1 AUTHOR

Cosimo Streppone, cosimo@cpan.org

=head1 SEE ALSO

L<Device::Gsm>, perl(1)

=cut
